pkg/controllers/execution/execution_controller.go

```diff
@@ -358,6 +358,16 @@ func (c *Controller) updateWorkDispatchingConditionIfNeeded(ctx context.Context,
 		return err
 	}
 
+	if work.GetObjectKind().GroupVersionKind().Empty() {
+		// In unit tests, we need to verify how many events are emitted. However, when using the fake client,
+		// the work object may not have TypeMeta (APIVersion/Kind) set, which prevents building a proper
+		// object reference and thus no event is actually emitted. This workaround ensures the GVK is set
+		// so event emission, and thereby event counting in unit tests, works as expected.
+		// Since controller-runtime v0.22.0, the group version kind is not set when using fake client.
+		// See https://github.com/kubernetes-sigs/controller-runtime/pull/3229 for more details.
+		work.SetGroupVersionKind(workv1alpha1.SchemeGroupVersion.WithKind("Work"))
+	}
```

Comment:
@zach593 Any thoughts? Can we do it this way?
The controller implementation relies on the `TypeMeta(APIVersion/Kind)`, and it is properly set when fetching the object with `client.Client`. 

But the unit test uses the `fake client`, the `TypeMeta` is removed, so the test won't get the expected event.

Reply:
Firstly, the `helper.ToUnstructured()` is still part of the controller; secondly, it's not that convenient to set the GVK as that will require an additional `rest mapper` to get the GVK dynamically.

Reply:
@zach593 Kindly ping~